source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -301,11 +302,20 @@ void SslSocket::shutdownSsl() {
   }
 }
 
+SslSocketInfo::SslSocketInfo(bssl::UniquePtr<SSL> ssl, ContextImplSharedPtr ctx)
+    : ssl_(std::move(ssl)) {
+  SSL_set_ex_data(ssl_.get(), ctx->sslCustomDataIndex(), &(this->certificate_validation_status_));
```

Comment:
Not thrilled to have more ex_data exposed into SSL library later for this. Can we consolidate this with existing `SSL_set_app_data` by defining a new interface for that purpose?

Reply:
Currently the set_app_data is on the ssl context for ContextImpl, and there exists a set_app_data on the SSL socket in the tls_inspector, so we weren't going to change that.
Because we're using this to be able to provide extra data related to the SSL connection, we've created an SslExtendedSocketInfo class that can be used/extended to pass this information through.
Let me know what you think of this approach.

Reply:
@lizan can you please give this change a re-review when you get a chance, thanks.

Reply:
> Currently the set_app_data is on the ssl context for ContextImpl, and there exists a set_app_data on the SSL socket in the tls_inspector, so we weren't going to change that.
> Because we're using this to be able to provide extra data related to the SSL connection, we've created an SslExtendedSocketInfo class that can be used/extended to pass this information through.
> Let me know what you think of this approach (included in the latest commits).

The `SSL_CTX_set_app_data` set `ContextImpl` which is not in the same scope, and TLS inspector `SSL` object is not leaked into here.

The only existing `SSL_set_app_data` is also to provide extra data related to SSL connection for transport socket options [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/transport_sockets/tls/context_impl.cc#L831), I meant to refactor these two into one interface and do `SSL_set_app_data` once.

Reply:
Could we do this refactor as a subsequent PR - I'm happy to pick this up.
because this may be a bit of a tricky refactor to combine the multiple pieces of info 'cleanly'.
we've got the ContextImpl's `const Network::TransportSocketOptions* options` being set via `SSL_set_app_data` by the ContextImpl but only `if (options && !options->verifySubjectAltNameListOverride().empty())` when the SslSocket is created,
and then the `SslSocketInfo`::`SslExtendedSocketInfo` member being set via `SSL_set_ex_data` when the `SslSocket` is subsequently moved to be owned by the constructed `SslSocketInfo`.

Reply:
@lizan Can we re-open this PR please - or do we need to create another PR

Reply:
> Could we do this refactor as a subsequent PR - I'm happy to pick this up.
> because this may be a bit of a tricky refactor to combine the multiple pieces of info 'cleanly'.
> we've got the ContextImpl's `const Network::TransportSocketOptions* options` being set via `SSL_set_app_data` by the ContextImpl but only `if (options && !options->verifySubjectAltNameListOverride().empty())` when the SslSocket is created,
> and then the `SslSocketInfo`::`SslExtendedSocketInfo` member being set via `SSL_set_ex_data` when the `SslSocket` is subsequently moved to be owned by the constructed `SslSocketInfo`.

Since this PR is relatively large, so sure, can you file an issue for this and assign yourself?

Reply:
Thanks, I've created issue: #9867 , but don't appear to have the power to assign myself - can you assign it to me please.